### ISSUE TITLE
Refactoring: Make method async and update callsites

### ISSUE THREAD

**Issue by alrz** [microsoft/vs-threading#746]
#### Is your feature request related to a problem? Please describe.

Migrating a codebase to use async overloads where available and make the enclosing methods async as well.

#### Describe the solution you'd like

Whenever an async overload is available **in a non-async** method, suggest to use that, make method async and update call-sites.

If this is done as a fix-all, this can be recursively applied to the call graph.

#### Describe alternatives you've considered

Currently this can be done in a few repetitive steps until the whole thing is migrated, but it is extremely cumbersome:

1. Change some method call to Async overload to get VSTHRD110, or
Change method return type to Task to get VSTHRD103
2. Use 'Add await' codefix to get CS4033
3. Use 'Make method async' to get VSTHRD110 or CS4014 on call-sites
4. Repeat

**Comment by alrz** [comment]
There is a open request for this here: https://github.com/dotnet/roslyn/issues/36737 but I figured all the necessary infra is implemented in this repo.

**Comment by AArnott** [comment]
My memory is unclear, but at least one of our code fixes may already do this:
https://github.com/microsoft/vs-threading/blob/d1dd01020458585e5f7f65aa044ca5f27f5982f8/src/Microsoft.VisualStudio.Threading.Analyzers.CodeFixes/FixUtils.cs#L78

That method will not only make a method async but also update call sites to await it. It *may* do it recursively but if not, at least it may reduce the steps you listed above. I just can't remember the conditions that lead to offering that code fix.

**Comment by alrz** [comment]
Yes it certainly *can* but it's not offered when the method is not returning `Task`, so the steps above need to be done to trigger it.

**Comment by alrz** [comment]
I updated the title to clarify. Note that awaiting call-sites means that we still need to run "make async and update call-sites" on the **containing methods** and this goes on until we get to a method that is already async.

**Comment by AArnott** [comment]
Ok. Ya, that would be great. 
I suspect other priorities will prevent anyone around my team from satisfying this request in the foreseeable future. If you're interested in sending a PR that includes several tests to demonstrate its capabilities and limitations, I'd be willing to review and possibly merge it.

**Comment by Bosch-Eli-Black** [comment]
Would _love_ to see this.

**Comment by alrz** [comment]
>I suspect other priorities will prevent anyone around my team from satisfying this request in the foreseeable future. If you're interested in sending a PR that includes several tests to demonstrate its capabilities and limitations, I'd be willing to review and possibly merge it.

Well I'm on the fence between this approach versus a code generator for my use case. We share the BL between two projects (aspnetcore API and a legacy website). A code generator could generate Async overloads where there are Async alternatives available in the body, then we can migrate only the API to use Async without touching the legacy code. Eitherway this is not a trivial task. 😢
